File: programmers/stack_queue/기능개발.py
from collections import deque


def my_solution(progresses, speeds):
    queue = deque(zip(progresses, speeds))
    answer = []

    while queue:
        queue = deque(map(lambda x: (x[0]+x[1], x[1]), queue))

        temp = 0
        while queue and queue[0][0] >= 100:
            queue.popleft()
            temp += 1

        if temp:
            answer.append(temp)

    return answer


def best_solution(progresses, speeds):
    '''
    1.
    내 방법은 모든 과정이 100이 넘을 때까지 계속 반복하고
    매번 deque를 재생성하지만,
    (speeds와 progresses 수치가 낮을수록 시간 증가)

    여기선 남은 일수의 대소 비교를 통해 전체 일 갯수 만큼만
    돌면 되서 아주 빠르며 굳이 queue를 안써도 된다.
    (두 수치가 크든 작든 상관 없음)

    2.
    math.ceil((100-p)//s) 대신에 -((p-100)//s)를 사용하여
    자동으로 올림을 사용
    '''
    Q = []
    for p, s in zip(progresses, speeds):
        if len(Q) == 0 or Q[-1][0]<-((p-100)//s):
            Q.append([-((p-100)//s), 1])
        else:
            Q[-1][1] += 1
    return [q[1] for q in Q]


def new_solution(progresses, speeds):
    '''
    best solution을 기억하며 다시 작성한 코드
    '''
    answer = []
    
    for p, s in zip(progresses,speeds):
        remain_day = -((p-100)//s)
        
        if not answer or remain_day > answer[-1][0]:
            answer.append([remain_day, 1])
        else:
            answer[-1][1] += 1

    return [x[1] for x in answer]

File: programmers/algorithm_data_structure/stack_queue/기능개발.java
package algorithm_data_structure.stack_queue;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class 기능개발 {

    public int[] mySolution(int[] progresses, int[] speeds) {
        List<Integer> answer = new ArrayList<>();
      
        int currTurn = 0;
        for (int i = 0; i < progresses.length; i++){
            int days = (int) Math.ceil((100.0 - progresses[i]) / speeds[i]);

            if (days <= currTurn){
                answer.set(answer.size()-1, answer.get(answer.size()-1)+1);                
            }else{
                answer.add(1);
                currTurn = days;
            }
        }

        return answer.stream().mapToInt(i -> i).toArray();
    }

    /*
     * 매번 완료일자를 찾아 그 index에 1을 더한다.
     * 반환할 때 배열 내 0이 아닌 index의 값을 반환한다.
     * 
     * 좋은 코드라기보단, 이렇게 풀 수 있다 정도로 이해
     */
    public int[] bestSolution(int[] progresses, int[] speeds){
        int[] dayOfEnd = new int[100];
        int currTurn = 0;
        for(int i = 0; i < progresses.length; i++){
            int days = (int) Math.ceil((100.0 - progresses[i]) / speeds[i]);
            if (days > currTurn){
                currTurn = days;
            }
            dayOfEnd[currTurn-1] += 1;
        }

        return Arrays.stream(dayOfEnd).filter(i -> i != 0).toArray();
    }
}